pkg/resource/deploy/source_eval.go

```diff
@@ -837,36 +870,67 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 	logging.V(5).Infof(
 		"ResourceMonitor.RegisterResource received: t=%v, name=%v, custom=%v, #props=%v, parent=%v, protect=%v, "+
 			"provider=%v, deps=%v, deleteBeforeReplace=%v, ignoreChanges=%v, aliases=%v, customTimeouts=%v",
-		t, name, custom, len(props), parent, protect, provider, dependencies, deleteBeforeReplace, ignoreChanges,
+		t, name, custom, len(props), parent, protect, providerRef, dependencies, deleteBeforeReplace, ignoreChanges,
 		aliases, timeouts)
 
-	// Send the goal state to the engine.
-	step := &registerResourceEvent{
-		goal: resource.NewGoal(t, name, custom, props, parent, protect, dependencies, provider, nil,
-			propertyDependencies, deleteBeforeReplace, ignoreChanges, additionalSecretOutputs, aliases, id, &timeouts),
-		done: make(chan *RegisterResult),
-	}
+	// If this is a remote component, fetch its provider and issue the construct call. Otherwise, register the resource.
+	var result *RegisterResult
+	var outputDeps map[string]*pulumirpc.RegisterResourceResponse_PropertyDependencies
+	if remote {
+		provider, ok := rm.providers.GetProvider(providerRef)
+		if !ok {
+			return nil, errors.Errorf("unknown provider '%v'", providerRef)
+		}
 
-	select {
-	case rm.regChan <- step:
-	case <-rm.cancel:
-		logging.V(5).Infof("ResourceMonitor.RegisterResource operation canceled, name=%s", name)
-		return nil, rpcerror.New(codes.Unavailable, "resource monitor shut down while sending resource registration")
-	}
+		// Invoke the provider's Construct RPC method.
+		options := plugin.ConstructOptions{
+			Aliases:              aliases,
+			Protect:              protect,
+			PropertyDependencies: propertyDependencies,
+		}
+		constructResult, err := provider.Construct(rm.constructInfo, t, name, parent, props, options)
+		if err != nil {
+			return nil, err
+		}
+		result = &RegisterResult{State: &resource.State{URN: constructResult.URN, Outputs: constructResult.Outputs}}
 
-	// Now block waiting for the operation to finish.
-	var result *RegisterResult
-	select {
-	case result = <-step.done:
-	case <-rm.cancel:
-		logging.V(5).Infof("ResourceMonitor.RegisterResource operation canceled, name=%s", name)
-		return nil, rpcerror.New(codes.Unavailable, "resource monitor shut down while waiting on step's done channel")
+		outputDeps = map[string]*pulumirpc.RegisterResourceResponse_PropertyDependencies{}
+		for k, deps := range constructResult.OutputDependencies {
+			urns := make([]string, len(deps))
+			for i, d := range deps {
+				urns[i] = string(d)
+			}
+			outputDeps[string(k)] = &pulumirpc.RegisterResourceResponse_PropertyDependencies{Urns: urns}
+		}
+	} else {
+		// Send the goal state to the engine.
+		step := &registerResourceEvent{
+			goal: resource.NewGoal(t, name, custom, goalProps, parent, protect, dependencies,
```

Comment:
Side note: some of these function signatures are getting rather long. It might be worth refactoring them to use an arg struct at some point.

Reply:
All these functions do is build a struct, so we could just use the struct itself :)